### PR TITLE
Add SAML Single Logout support

### PR DESCRIPTION
**Pull request by larsven**
We think it would be great to have support for SAML Single Logout so we added this.

Please review and let us know if you would like us to make any changes. Thank you!

### CODE REVIEW

File: src/main/java/eu/openanalytics/containerproxy/auth/impl/SAMLAuthenticationBackend.java [hunk 1]
```diff
@@ -67,6 +67,9 @@ public void configureHttpSecurity(HttpSecurity http) throws Exception {
 		.and()
 			.addFilterBefore(metadataGeneratorFilter, ChannelProcessingFilter.class)
 			.addFilterAfter(samlFilter, BasicAuthenticationFilter.class);
+		http
+				.logout()
+				.disable();
```
Comment (LEDfan) [inline src/main/java/eu/openanalytics/containerproxy/auth/impl/SAMLAuthenticationBackend.java:72]: Why is this change required? From my tests, it breaks the logout button.
Even when using Single Logout, the user should be able to logout, right?

**Comment by LEDfan** [comment]
Hi @larsven 

Thanks for your contribution! We're currently looking to integrate this feature.
To ensure no misunderstandings, this PR adds support for the following scenario:
 - two applications connected to the same SAML server, application x and ShinyProxy
 - somebody logs out from application x
 - the SAML server sends a request to ShinyProxy to logout the user from ShinyProxy

Is that correct?

Another question: which SAML server/implementation did you use to test this?
We usually use Keycloak, but apparently it has no support for Single Logout.

I left a comment on the code, please have a look at that.

**Comment by larsven** [comment]
@LEDfan 

> Is that correct?

A more common use case is that when someone logs out from ShinyProxy (the SAML Service Provider in this case), they will also be logged out from the SAML Identity Provider (for example KeyCloak). If fully propagated logout is implemented and enabled in the SAML Identity Provider, the user will also be logged out from “application x” (like you describe above), but not all SAML Identity Provider softwares/services support fully propagated logout.

> Another question: which SAML server/implementation did you use to test this?

We used Okta to test this. KeyCloak does support SAML Single Logout and according to its documentation it also supports the fully propagated logout use case, but we have not tested this.

> I left a comment on the code, please have a look at that.

These 3 lines are no longer needed and we have updated our branch to reflect this.

Additionally:

- We improved the PR to provide a more generic solution to support the logout button. Logout buttons located in the navigation bar and in the container app now support logout for SAML and other authentication methods.
- ContainerProxy 0.8.6 version introduces a runtime error related to SAML, Kerberos and possibly other authentication methods. This bug was fixed in the “develop” branch, but has NOT yet been merged into the master branch. SAML will not work without this fix: https://github.com/openanalytics/containerproxy/blob/develop/src/main/java/eu/openanalytics/containerproxy/security/WebSecurityConfig.java#L169-L174

**Comment by LEDfan** [comment]
Hi @larsven 

> A more common use case is that when someone logs out from ShinyProxy (the SAML Service Provider in this case), they will also be logged out from the SAML Identity Provider (for example KeyCloak)

Okay, that's interesting, because I believe we already support this. Since [ShinyProxy 2.4.0](https://shinyproxy.io/downloads/#240) you can specify the `proxy.saml.logout-url` property. When specified the logout button is changed so that you are redirected to that URL. This is the PR that included this feature: https://github.com/openanalytics/containerproxy/pull/27/files.
In our tests, we use it with Auth0 and that works fine.

Could you test whether settings this property for you?

> ContainerProxy 0.8.6 version introduces a runtime error related to SAML, Kerberos and possibly other authentication methods. This bug was fixed in the “develop” branch, but has NOT yet been merged into the master branch. 

Indeed, there was an issue with that release. Last Friday we releases [ContainerProxy 0.8.7](https://github.com/openanalytics/containerproxy/releases/tag/v0.8.7 and [ShinyProxy 2.4.3](https://shinyproxy.io/downloads/#243) with fixes for this issue.

**Comment by larsven** [comment]
Hi @LEDfan,

Yes, we saw that property. The getLogoutSuccessURL() method that you used to logout is usually used to redirect the user to the specified URL after the logout process has successfully completed. This method does a redirect without any query parameters. This is not how it is supposed to work according to the SAML specification; a SAML logout request must contain a SAML logout payload. Some Identity Providers, like for example Okta, even require the logout request to be signed.

We tested your approach with Okta and as expected it did not work. We also tested with Auth0 but could not get that to work either - are you using some specific configuration to disable the need for a SAML logout payload to be sent?

**Comment by LEDfan** [comment]
Hi

Thanks for explaining this. I just had a look at the Auth0 documentation. In our test configuration we are using the [/logout endpoint](https://auth0.com/docs/api/authentication#logout). This is basically a special API of Auth0 to logout users. Therefore it is not part of the SAML specification and therefore it does not require the SAML logout payload. I wasn't aware of this, until you explained this.

Therefore, I see the value of this PR and I'm going to test this in the coming days (or maybe next week). Of course, a correct implementation of SAML is very important to us.

**Comment by LEDfan** [comment]
I tested the implementation and it works as expected, therefore I pushed you changes into our develop branch (see https://github.com/openanalytics/containerproxy/commits/develop) . I did make some changes, so that we don't need the ` LogoutController`. 
Furthermore, I also took the opportunity to make the necessary changes so that SAML Single Logout works, using the `/saml/SingleLogout` endpoint. 

Would you please be so kind to test these changes on you infrastructure?

I'm closing this PR, because we added the commits using another way, nevertheless you should be credited for the commits. Thanks gain for your contribution!

**Comment by LEDfan** [comment]
FYI you also need the latest commit of ShinyProxy before this will work (https://github.com/openanalytics/shinyproxy/commit/289be818302601e7b418cc1b556d4463aaeec96d)

**Comment by larsven** [comment]
Hi @LEDfan,

We have tested SAML SLO with these versions:
https://github.com/openanalytics/containerproxy/commits/develop
https://github.com/openanalytics/shinyproxy/commit/289be818302601e7b418cc1b556d4463aaeec96d

In a simple integration between one ShinyProxy application and Okta, logout worked well.

But when we tested with two different ShinyProxy integrations at the same time, it didn’t work as expected. When we are logged in to two different ShinyProxy applications we have to press the logout button in the first application twice to get logged out. This behavior cannot be reproduced when testing with our commit of ContainerProxy and the release version of ShinyProxy.

**Comment by LEDfan** [comment]
Hi @larsven 

Thanks for testing, do you perhaps have the `proxy.saml.logout-url` property set?  In our current implementation, we check whether this property is set, if so we don't use the `/saml/logout` URL. This could cause what you are describing. However, note that we are probably going to reconsider this behavior.

Thanks

**Comment by larsven** [comment]
Hi @LEDfan,

We tested it with Okta which requires SAML logout payload. So we tested it without setting the “proxy.saml.logout-url” property.

**Comment by LEDfan** [comment]
Hi @larsven 

I looked into this yesterday and setup two applications inside Okta and configured these in ShinyProxy. I could reproduce the problem you were describing, but only in a very specific situation. That is, only when I'm running two ShinyProxy servers (configured with Okta) on the same domain but with a different port.
For example, I tested with:

 - `localhost:8080`
 - `localhost:9070`

This is what happens:

1. user goes to `localhost:8080` and logs in with Okta, the user now gets a session cookie with the name `JSESSIONID` and value `xyz123`
2. in the same browser (but a different tab) the user goes to `localhost:9070` and gets automatically logged in using Okta. ShinyProxy now overrides the cookie `JSESSIONID` with value `abc`. Note that cookies are not tight to a port, only to the hostname (and optionally the path)
3. user goes to the tab with `localhost:8080` and clicks the signout button. This ShinyProxy server does not recognize the cookie with value `abc`, so the user is redirected to Okta and gets automatically logged in. 
4. The user again clicks the signout button. The user now again has a proper session cookie and the logout now works properly.
5. The user refreshes the tab with `localhost:9070` and is logged out because the `JSESSIONID` is destroyed.

When I configure two ShinyProxies with a different domain name, everything works as expected.
So my question is, were you testing this with the same domain name (e.g. localhost), but a different port?

P.s. even with your branch I could reproduce the issue when using `localhost` and different ports.

**Comment by larsven** [comment]
Hi @LEDfan,

> So my question is, were you testing this with the same domain name (e.g. localhost), but a different port?

Yes, that's what we did. But from our point of view this should be supported.

> P.s. even with your branch I could reproduce the issue when using localhost and different ports.

To us it works. We tested again today. Probably there is some other difference between our environments.

**Comment by LEDfan** [comment]
Hi @larsven 

Thanks for your feedback!
It is very strange that this is working on your end, as it should not work properly. Even if you use the `simple` authentication backend all sort of weird things will happen if you run ShinyProxy on the same domain and path but with a different port. The session cookie will be shared between multiple servers but they will think that they have sole ownership of the cookie. Even with simple authentication things like session fixation, logout etc will not work properly, simply because each server is overwriting the cookie. However, to trigger the bug (i.e. the overwriting of the cookie), it's import to reload the page. For example before logging out on ShinyProxy server 1, make sure to reload the page on ShinyProxy 2 (this will make sure that ShinyProxy 1 doesn't recognize the cookie).

Therefore, I believe this is unrelated to any change in the SAML code.

> But from our point of view this should be supported.

If you want to use such a setup (which I agree may be a valid use-case), you should make sure each ShinyProxy server uses a unique cookie for their session.
For example using:
 1. don't change anything in the configuration of ShinyProxy Server 1
 2. add the following config to ShinyProxy 2:
```
server:
  servlet:
    session:
      cookie:
        name: JSESSIONID2
```

AFAIK this is the only way to make such a setup work. Of course I could be wrong somewhere here, but in my testes and observations this is the only way to make it work properly.

Could you try with this configuration change and report back whether everything works properly?

**Comment by LEDfan** [comment]
Hi @larsven 

We just released ShinyProxy 2.5.0 which now includes your PR. I mentioned https://github.com/DefensePoint in our release notes (https://shinyproxy.io/downloads/#250). Please let me know if that's not the proper account or if you prefer to not be mentioned.

Thanks again for your contribution!
